planner/core/point_get_plan.go

```diff
@@ -184,96 +261,216 @@ func TryFastPlan(ctx sessionctx.Context, node ast.Node) Plan {
-	if selStmt.OrderBy != nil || selStmt.GroupBy != nil || selStmt.Limit != nil ||
-		selStmt.Having != nil || len(selStmt.WindowSpecs) > 0 ||
-		selStmt.LockTp != ast.SelectLockNone {
+	if selStmt.OrderBy != nil || selStmt.GroupBy != nil ||
+		selStmt.Limit != nil || selStmt.Having != nil ||
+		len(selStmt.WindowSpecs) > 0 || selStmt.LockTp != ast.SelectLockNone {
 		return nil
 	}
 	in, ok := selStmt.Where.(*ast.PatternInExpr)
 	if !ok || in.Not || len(in.List) < 1 {
 		return nil
 	}
 
-	children := make([]PhysicalPlan, 0, len(in.List))
-	chReqProps := make([]*property.PhysicalProperty, 0, len(in.List))
-	reusedStmt := &ast.SelectStmt{
-		SelectStmtOpts: selStmt.SelectStmtOpts,
-		Distinct:       selStmt.Distinct,
-		From:           selStmt.From,
-		Fields:         selStmt.Fields,
+	tblName, tblAlias := getSingleTableNameAndAlias(selStmt.From)
+	if tblName == nil {
+		return nil
+	}
+	tbl := tblName.TableInfo
+	if tbl == nil {
+		return nil
+	}
+
+	// Do not handle partitioned table.
+	// Table partition implementation translates LogicalPlan from `DataSource` to
+	// `Union -> DataSource` in the logical plan optimization pass, since BatchPointGetPlan
+	// bypass the logical plan optimization, it can't support partitioned table.
+	if tbl.GetPartitionInfo() != nil {
+		return nil
+	}
+
+	for _, col := range tbl.Columns {
+		if col.IsGenerated() || col.State != model.StatePublic {
+			return nil
+		}
+	}
+
+	schema, names := buildSchemaFromFields(tblName.Schema, tbl, tblAlias, selStmt.Fields.Fields)
+	if schema == nil {
+		return nil
 	}
 
-	switch leftExpr := in.Expr.(type) {
+	var (
+		tryHandle     bool
+		fieldType     *types.FieldType
+		whereColNames []string
+	)
+
+	// SELECT * FROM t WHERE (key) in ((1), (2))
+	colExpr := in.Expr
+	if p, ok := colExpr.(*ast.ParenthesesExpr); ok {
+		colExpr = p.Expr
+	}
+	switch colName := colExpr.(type) {
 	case *ast.ColumnNameExpr:
-		reusedStmt := &ast.SelectStmt{
-			SelectStmtOpts: selStmt.SelectStmtOpts,
-			Distinct:       selStmt.Distinct,
-			From:           selStmt.From,
-			Fields:         selStmt.Fields,
-		}
-		for _, row := range in.List {
-			where := &ast.BinaryOperationExpr{
-				Op: opcode.EQ,
-				L:  in.Expr,
-				R:  row,
-			}
-			reusedStmt.Where = where
-			fp := TryFastPlan(ctx, reusedStmt)
-			if fp == nil {
-				return nil
+		if name := colName.Name.Table.L; name != "" && name != tblAlias.L {
+			return nil
+		}
+		// Try use handle
+		if tbl.PKIsHandle {
+			for _, col := range tbl.Columns {
+				if mysql.HasPriKeyFlag(col.Flag) {
+					tryHandle = col.Name.L == colName.Name.Name.L
+					fieldType = &col.FieldType
+					whereColNames = append(whereColNames, col.Name.L)
+					break
+				}
 			}
-			chReqProps = append(chReqProps, &property.PhysicalProperty{ExpectedCnt: 1})
-			children = append(children, fp.(*PointGetPlan))
 		}
 
 	case *ast.RowExpr:
-		if len(leftExpr.Values) < 1 {
+		if len(colName.Values) < 1 {
```

Comment:
When would this happen?

Reply:
Seems never happen, I will remove this part.